src/rulesets/oas2/functions/oas2UnusedDefinition.ts

```diff
@@ -0,0 +1,58 @@
+import { JsonPath } from '@stoplight/types';
+import { IFunction, IFunctionResult } from '../../../types';
+
+function* refIterator(obj: object, path: JsonPath): IterableIterator<[JsonPath, string]> {
+  for (const key in obj) {
+    if (!Object.hasOwnProperty.call(obj, key)) continue;
+    const scopedPath = [...path, key];
+    if (key === '$ref') {
+      yield [scopedPath, obj[key]];
+    } else if (typeof obj[key] === 'object' && obj[key] !== null) {
+      yield* refIterator(obj[key], scopedPath);
+    }
+  }
+}
+
+interface IDescriptor {
+  name: string;
+  seen: boolean;
+}
+
+const oas2UnusedDefinition: IFunction = data => {
+  const results: IFunctionResult[] = [];
+
+  if (typeof data !== 'object' || data === null) return results;
+
+  if (!data.definitions) {
+    return results;
+  }
+
+  const seen: Record<string, IDescriptor> = {};
+
+  Object.keys(data.definitions).forEach(defName => {
+    seen[`#/definitions/${defName}`] = { name: defName, seen: false };
+  });
+
+  const declaredDefinitions = Object.keys(seen);
+
+  for (const [, refContent] of refIterator(data, [])) {
+    if (!declaredDefinitions.includes(refContent)) {
+      continue;
```

Comment:
Note to self: Add a reference pointing to an external object to an existing test to cover this.